File: android/app/src/main/java/app/releaf/mobile/ui/components/editor/OverviewPane.kt
/*
 * OverviewPane.kt
 *
 * Alternative layout for the notepad and notebook-page editors — the
 * "Overview" side of the EditorModeToggle. Looks like the
 * `PageDetailView` (home → notebook → page) on purpose: same
 * `CaptureTabBar` + section-per-tab structure, but fed by the editor's
 * own mutable state so add/remove + body edits still persist through
 * the VM.
 *
 * Editable Overview tab: the notes body binds to the same `RichTextState`
 * Edit mode uses, so a change in either mode shows up live in the
 * other. The formatting toolbar pins to the bottom of the pane but is
 * only visible while the Overview tab is selected — the other tabs
 * manage plain-text / list / grid interactions that don't need inline
 * formatting.
 *
 * All seven capture modes are surfaced now — Voice was re-enabled once
 * `VoiceSection` landed. Each tab gets its own section body; Overview
 * shows the stat grid + notes preview card.
 */

package app.releaf.mobile.ui.components.editor

import android.net.Uri
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Edit
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.saveable.Saver
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.unit.dp
import app.releaf.mobile.data.notebook.Attachment
import app.releaf.mobile.data.notebook.Contact
import app.releaf.mobile.data.notebook.GeoLocation
import app.releaf.mobile.data.notebook.TodoItem
import app.releaf.mobile.ui.components.CaptureMode
import app.releaf.mobile.ui.components.CaptureTabBar
import app.releaf.mobile.ui.components.StatGrid
import app.releaf.mobile.ui.components.StatItem
import app.releaf.mobile.ui.components.StatTone
import androidx.compose.ui.unit.sp
import app.releaf.mobile.ui.theme.AppColors
import app.releaf.mobile.ui.theme.AppRadius
import app.releaf.mobile.ui.theme.AppSpacing
import app.releaf.mobile.ui.theme.AppTypography
import com.mikepenz.markdown.m3.Markdown
import com.mohamedrejeb.richeditor.model.RichTextState

/**
 * Overview-mode editor pane. `CaptureTabBar` + per-tab scrolling
 * content. Title, eyebrow, and date chip all live at screen level
 * now (above this pane), so the pane itself only owns the tab row
 * and its content.
 */
@Composable
fun OverviewPane(
    richTextState: RichTextState,
    contacts: List<Contact>,
    todos: List<TodoItem>,
    locations: List<GeoLocation>,
    attachments: List<Attachment>,
    onAddContact: (String) -> Unit,
    onRemoveContact: (String) -> Unit,
    onAddTodo: (String) -> Unit,
    onToggleTodo: (String) -> Unit,
    onRemoveTodo: (String) -> Unit,
    onAddLocation: (Double, Double, String?) -> Unit,
    onRemoveLocation: (String) -> Unit,
    onAddPhoto: (String) -> Unit,
    onAddScan: (String, String?, List<Uri>) -> Unit,
    onAddVoiceNote: (String, Long) -> Unit,
    onTranscribeVoiceNote: (String, String?) -> Unit,
    onRemoveAttachment: (String) -> Unit,
    modifier: Modifier = Modifier,
) {
    // All seven modes render; Voice got its own section once
    // `VoiceSection` landed. The tab row takes `CaptureMode.entries`
    // directly.
    val modes = remember { CaptureMode.entries.toList() }

    var selected by rememberSaveable(stateSaver = captureModeSaver) {
        mutableStateOf(CaptureMode.Overview)
    }
    var notesSheetOpen by rememberSaveable { mutableStateOf(false) }

    Column(modifier = modifier.fillMaxSize()) {
        // --- CaptureTabBar ---
        CaptureTabBar(
            selected = selected,
            onSelect = { selected = it },
            modes    = modes,
        )

        // --- Tab body ---
        Box(Modifier.weight(1f)) {
            Column(
                modifier = Modifier
                    .fillMaxSize()
                    .verticalScroll(rememberScrollState())
                    .padding(AppSpacing.s4),
                verticalArrangement = Arrangement.spacedBy(AppSpacing.s4),
            ) {
                when (selected) {
                    CaptureMode.Overview -> OverviewTab(
                        richTextState = richTextState,
                        contacts      = contacts,
                        todos         = todos,
                        locations     = locations,
                        attachments   = attachments,
                        onEditNotes   = { notesSheetOpen = true },
                    )
                    CaptureMode.Photos -> PhotosSection(
                        photos   = attachments.filter { it.type == Attachment.TYPE_PHOTO },
                        onAdd    = onAddPhoto,
                        onRemove = onRemoveAttachment,
                    )
                    CaptureMode.Scans -> ScansSection(
                        scans    = attachments.filter { it.type == Attachment.TYPE_SCAN },
                        onAdd    = onAddScan,
                        onRemove = onRemoveAttachment,
                    )
                    CaptureMode.Voice -> VoiceSection(
                        notes         = attachments.filter { it.type == Attachment.TYPE_VOICE },
                        onAdd         = onAddVoiceNote,
                        onTranscribed = onTranscribeVoiceNote,
                        onRemove      = onRemoveAttachment,
                    )
                    CaptureMode.Todo -> TodosSection(
                        todos    = todos,
                        onAdd    = onAddTodo,
                        onToggle = onToggleTodo,
                        onRemove = onRemoveTodo,
                    )
                    CaptureMode.Contacts -> ContactsSection(
                        contacts = contacts,
                        onAdd    = onAddContact,
                        onRemove = onRemoveContact,
                    )
                    CaptureMode.Location -> LocationSection(
                        locations = locations,
                        onAdd     = onAddLocation,
                        onRemove  = onRemoveLocation,
                    )
                }
                Spacer(Modifier.height(AppSpacing.s10))
            }
        }

    }

    // Full-screen editing sheet. Binds to the same RichTextState, so
    // anything typed here flows back into the Overview preview on
    // dismiss and into Edit mode automatically.
    if (notesSheetOpen) {
        NotesEditorSheet(
            richTextState = richTextState,
            onDismiss     = { notesSheetOpen = false },
        )
    }
}

/**
 * Overview-tab content: at-a-glance stat grid + tappable notes card.
 * Tapping the card opens the full-screen `NotesEditorSheet` — the
 * inline editor-in-ScrollView pattern proved too cramped alongside
 * the stats. Counts still update live from the current
 * `RichTextState` so the "Words" stat reflects the live buffer.
 */
@Composable
private fun OverviewTab(
    richTextState: RichTextState,
    contacts: List<Contact>,
    todos: List<TodoItem>,
    locations: List<GeoLocation>,
    attachments: List<Attachment>,
    onEditNotes: () -> Unit,
) {
    val photos = attachments.count { it.type == Attachment.TYPE_PHOTO }
    val scans  = attachments.count { it.type == Attachment.TYPE_SCAN }
    val bodyText = richTextState.annotatedString.text
    val words  = if (bodyText.isBlank()) 0 else bodyText.trim().split(Regex("\\s+")).size

    Text(
        "AT A GLANCE",
        style = AppTypography.Eyebrow,
        color = AppColors.TextSecondary,
    )

    StatGrid(items = listOf(
        StatItem(label = "Photos",   value = "$photos",         tone = StatTone.Coral),
        StatItem(label = "Scans",    value = "$scans",          tone = StatTone.Neutral),
        StatItem(label = "To-do",    value = "${todos.size}",   tone = StatTone.Green),
    ))
    StatGrid(items = listOf(
        StatItem(label = "Contacts", value = "${contacts.size}",  tone = StatTone.Info),
        StatItem(label = "Places",   value = "${locations.size}", tone = StatTone.Neutral),
        StatItem(label = "Words",    value = "$words",            tone = StatTone.Neutral),
    ))

    NotesPreviewCard(
        richTextState = richTextState,
        onTap         = onEditNotes,
    )
}

/**
 * Tap target that shows a rendered-markdown preview of the current
 * notes body (or an empty-state call to action). Tapping fires
 * `onTap` which the caller uses to open `NotesEditorSheet`. A small
 * pencil icon in the corner emphasises the affordance.
 */
@Composable
private fun NotesPreviewCard(
    richTextState: RichTextState,
    onTap: () -> Unit,
) {
    // Subscribe explicitly to `annotatedString` so the preview
    // recomposes when the sheet or Edit-mode editor mutates the state.
    // `toMarkdown()` alone doesn't register a Compose state read,
    // which is why the preview showed stale content after the user
    // first dismissed the editing sheet.
    val annotated = richTextState.annotatedString
    val markdown = remember(annotated) { richTextState.toMarkdown() }
    val isEmpty = annotated.text.trim().isEmpty()

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(AppRadius.md))
            .background(AppColors.CardSolid)
            .border(
                width = 1.dp,
                color = AppColors.BorderDefault,
                shape = RoundedCornerShape(AppRadius.md),
            )
            .clickable(onClick = onTap)
            .padding(AppSpacing.s3),
        verticalArrangement = Arrangement.spacedBy(AppSpacing.s2),
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Text(
                "NOTES",
                style = AppTypography.Eyebrow,
                color = AppColors.TextSecondary,
            )
            Spacer(Modifier.weight(1f))
            Icon(
                imageVector        = Icons.Filled.Edit,
                contentDescription = "Edit notes",
                tint               = AppColors.Coral,
                modifier           = Modifier.size(18.dp),
            )
        }

        if (isEmpty) {
            Text(
                "Tap to write notes…",
                style = AppTypography.Body,
                color = AppColors.TextTertiary,
            )
        } else {
            Markdown(content = markdown)
        }
    }
}

// `OverviewTitleField` was removed — the title now lives in a row at
// screen level, shared between Edit and Overview modes. Keeping it
// here would duplicate the title surface when the user switches modes.

private val captureModeSaver: Saver<CaptureMode, String> = Saver(
    save    = { it.name },
    restore = { name -> runCatching { CaptureMode.valueOf(name) }.getOrDefault(CaptureMode.Overview) },
)

File: apps/releaf/ios/Releaf/Features/Notepad/OverviewPane.swift
/*
 * OverviewPane.swift
 *
 * Alternative notepad-editor layout matching `PageDetailView`
 * (home → notebook → page). Same `CaptureTabBar` + section-per-tab
 * shape, but fed by the editor's own mutable state so add/remove
 * routes through the VM.
 *
 * Notes editing moved off the inline canvas and into
 * `NotesEditorSheet` — tap the notes preview card to open a
 * dedicated full-sheet editor with the format toolbar pinned above
 * the keyboard. Keeps the Overview tab readable and still makes
 * editing a one-tap affordance.
 *
 * All seven capture modes are surfaced now — Voice was re-enabled
 * once `VoiceSection` landed. Each tab gets its own section body;
 * Overview shows the stat grid + notes preview card.
 */

import SwiftUI
import ReleafData
import ReleafDesignSystem

struct OverviewPane: View {
    @Binding var notes: String
    @ObservedObject var richTextController: RichTextEditorController
    let contacts: [NotepadContact]
    let todos: [NotepadTodo]
    let locations: [GeoLocation]
    let attachments: [Attachment]
    let onAddContact: (String) -> Void
    let onRemoveContact: (String) -> Void
    let onAddTodo: (String) -> Void
    let onToggleTodo: (String) -> Void
    let onRemoveTodo: (String) -> Void
    let onAddLocation: (Double, Double, String?) -> Void
    let onRemoveLocation: (String) -> Void
    let onAddPhoto: (String) -> Void
    let onAddScan: (String, String?) -> Void
    let onAddVoiceNote: (String, Int) -> Void
    let onTranscribeVoiceNote: (String, String?) -> Void
    let onRemoveAttachment: (String) -> Void
    /// Optional preselected CaptureTabBar tab. When non-nil, the pane
    /// opens with this tab active instead of the default `.overview`.
    /// Used by the Recents new-entry picker so tapping "Photo" lands
    /// the user on the Photos tab of the grid layout.
    var initialSelected: CaptureMode? = nil

    @State private var selected: CaptureMode
    @State private var notesSheetOpen: Bool = false

    init(
        notes: Binding<String>,
        richTextController: RichTextEditorController,
        contacts: [NotepadContact],
        todos: [NotepadTodo],
        locations: [GeoLocation],
        attachments: [Attachment],
        onAddContact: @escaping (String) -> Void,
        onRemoveContact: @escaping (String) -> Void,
        onAddTodo: @escaping (String) -> Void,
        onToggleTodo: @escaping (String) -> Void,
        onRemoveTodo: @escaping (String) -> Void,
        onAddLocation: @escaping (Double, Double, String?) -> Void,
        onRemoveLocation: @escaping (String) -> Void,
        onAddPhoto: @escaping (String) -> Void,
        onAddScan: @escaping (String, String?) -> Void,
        onAddVoiceNote: @escaping (String, Int) -> Void,
        onTranscribeVoiceNote: @escaping (String, String?) -> Void,
        onRemoveAttachment: @escaping (String) -> Void,
        initialSelected: CaptureMode? = nil
    ) {
        self._notes = notes
        self.richTextController = richTextController
        self.contacts = contacts
        self.todos = todos
        self.locations = locations
        self.attachments = attachments
        self.onAddContact = onAddContact
        self.onRemoveContact = onRemoveContact
        self.onAddTodo = onAddTodo
        self.onToggleTodo = onToggleTodo
        self.onRemoveTodo = onRemoveTodo
        self.onAddLocation = onAddLocation
        self.onRemoveLocation = onRemoveLocation
        self.onAddPhoto = onAddPhoto
        self.onAddScan = onAddScan
        self.onAddVoiceNote = onAddVoiceNote
        self.onTranscribeVoiceNote = onTranscribeVoiceNote
        self.onRemoveAttachment = onRemoveAttachment
        self.initialSelected = initialSelected
        // Seed the @State at init time so the route's mode hint lands
        // before the body renders.
        self._selected = State(initialValue: initialSelected ?? .overview)
    }

    /// All seven modes render — Voice was re-enabled once
    /// `VoiceSection` landed. Kept as a computed so we can drop modes
    /// again (e.g. for a future "read-only" flavor) without a second
    /// source of truth.
    private var modes: [CaptureMode] { CaptureMode.allCases }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            // Title + date row lives at screen level now — this pane
            // starts straight at the CaptureTabBar.
            CaptureTabBar(modes: modes, selected: $selected)

            ScrollView {
                VStack(alignment: .leading, spacing: AppSpacing.s4) {
                    sectionBody
                    Spacer(minLength: AppSpacing.s10)
                }
                .padding(AppSpacing.s4)
                .frame(maxWidth: .infinity, alignment: .leading)
            }
        }
        .sheet(isPresented: $notesSheetOpen) {
            NotesEditorSheet(
                notes:      $notes,
                controller: richTextController,
                onDismiss:  { notesSheetOpen = false }
            )
        }
    }

    // MARK: Tab body

    @ViewBuilder private var sectionBody: some View {
        switch selected {
        case .overview:
            OverviewTab(
                notes:       notes,
                contacts:    contacts,
                todos:       todos,
                locations:   locations,
                attachments: attachments,
                onEditNotes: { notesSheetOpen = true }
            )

        case .photos:
            PhotosSection(
                photos:   attachments.filter { $0.type == Attachment.typePhoto },
                onAdd:    onAddPhoto,
                onRemove: onRemoveAttachment
            )

        case .scans:
            ScansSection(
                scans:    attachments.filter { $0.type == Attachment.typeScan },
                onAdd:    onAddScan,
                onRemove: onRemoveAttachment
            )

        case .voice:
            VoiceSection(
                notes:         attachments.filter { $0.type == Attachment.typeVoice },
                onAdd:         onAddVoiceNote,
                onTranscribed: onTranscribeVoiceNote,
                onRemove:      onRemoveAttachment
            )

        case .todo:
            TodosSection(
                todos:    todos,
                onAdd:    onAddTodo,
                onToggle: onToggleTodo,
                onRemove: onRemoveTodo
            )

        case .contacts:
            ContactsSection(
                contacts: contacts,
                onAdd:    onAddContact,
                onRemove: onRemoveContact
            )

        case .location:
            LocationSection(
                locations: locations,
                onAdd:     onAddLocation,
                onRemove:  onRemoveLocation
            )

        // .notes is editor-only per CaptureMode.swift — selecting the
        // notes tab opens the editor sheet (`notesSheetOpen = true`),
        // not a section body. Placeholder satisfies exhaustive switch.
        case .notes:
            EmptyView()
        }
    }
}

// MARK: - Overview tab (at-a-glance + tappable notes card)

private struct OverviewTab: View {
    let notes: String
    let contacts: [NotepadContact]
    let todos: [NotepadTodo]
    let locations: [GeoLocation]
    let attachments: [Attachment]
    let onEditNotes: () -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: AppSpacing.s4) {
            Text("AT A GLANCE")
                .font(AppText.eyebrow)
                .tracking(AppLetterSpacing.eyebrow)
                .foregroundStyle(AppColors.textSecondary)

            let photosCount = attachments.filter { $0.type == Attachment.typePhoto }.count
            let scansCount  = attachments.filter { $0.type == Attachment.typeScan }.count

            StatGrid(items: [
                StatItem(label: "Photos", value: "\(photosCount)",  tone: .coral),
                StatItem(label: "Scans",  value: "\(scansCount)",   tone: .neutral),
                StatItem(label: "To-do",  value: "\(todos.count)",  tone: .green),
            ])
            StatGrid(items: [
                StatItem(label: "Contacts", value: "\(contacts.count)",  tone: .info),
                StatItem(label: "Places",   value: "\(locations.count)", tone: .neutral),
                StatItem(label: "Words",    value: "\(wordCount)",       tone: .neutral),
            ])

            NotesPreviewCard(notes: notes, onTap: onEditNotes)
        }
    }

    private var wordCount: Int {
        let trimmed = notes.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return 0 }
        return trimmed.split(whereSeparator: { $0.isWhitespace }).count
    }
}

/// Tappable card with a rendered-markdown preview. Pencil icon in the
/// corner signals the tap-to-edit affordance.
private struct NotesPreviewCard: View {
    let notes: String
    let onTap: () -> Void

    private var isEmpty: Bool {
        notes.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    private var attributedNotes: AttributedString {
        var options = AttributedString.MarkdownParsingOptions()
        options.interpretedSyntax = .inlineOnlyPreservingWhitespace
        return (try? AttributedString(markdown: notes, options: options))
            ?? AttributedString(notes)
    }

    var body: some View {
        Button(action: onTap) {
            VStack(alignment: .leading, spacing: AppSpacing.s2) {
                HStack {
                    Text("NOTES")
                        .font(AppText.eyebrow)
                        .tracking(AppLetterSpacing.eyebrow)
                        .foregroundStyle(AppColors.textSecondary)
                    Spacer()
                    Image(systemName: "pencil")
                        .font(.system(size: 14))
                        .foregroundStyle(AppColors.coral)
                }

                if isEmpty {
                    Text("Tap to write notes…")
                        .font(AppText.body)
                        .foregroundStyle(AppColors.textTertiary)
                } else {
                    Text(attributedNotes)
                        .font(AppText.body)
                        .foregroundStyle(AppColors.textPrimary)
                        .multilineTextAlignment(.leading)
                        .frame(maxWidth: .infinity, alignment: .leading)
                }
            }
            .padding(AppSpacing.s3)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(
                RoundedRectangle(cornerRadius: AppRadius.md, style: .continuous)
                    .fill(AppColors.cardSolid)
            )
            .overlay(
                RoundedRectangle(cornerRadius: AppRadius.md, style: .continuous)
                    .stroke(AppColors.borderDefault, lineWidth: 1)
            )
        }
        .buttonStyle(.plain)
    }
}
